### PR TITLE
UCP/API: nbx API for stream calls

### PR DESCRIPTION
**Pull request by hoopoepg**
- added nbx notation for stream API

### CODE REVIEW

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -2449,6 +2449,38 @@ ucs_status_ptr_t ucp_stream_send_nb(ucp_ep_h ep, const void *buffer, size_t coun
...
+ * it is safe to reuse the source @e buffer. If the send operation is
+ * completed immediately the routine returns UCS_OK.
+ *
+ * @note The user should not modify any part of the @a buffer after this
+ *       operation is called, until the operation completes.
+ *
+ * @param [in]  ep          Destination endpoint handle.
+ * @param [in]  buffer      Pointer to the message buffer (payload).
+ * @param [in]  count       Number of elements to send.
+ * @param [in]  flags       Reserved for future use.
```
Comment (brminich) [inline src/ucp/api/ucp.h:2469]: do we still need flags? Now we can add new param, right?
@evgeny-leksikov

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -2449,6 +2449,38 @@ ucs_status_ptr_t ucp_stream_send_nb(ucp_ep_h ep, const void *buffer, size_t coun
...
+ * @param [in]  param       Operation parameters, see @ref ucp_request_param_t
```
Comment (brminich) [inline src/ucp/api/ucp.h:2470]: missing `.` in the end

Comment (brminich) [inline src/ucp/api/ucp.h:2470]: can we specify `dt` param for this cal?

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -2732,6 +2764,39 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * operation cannot be started, the routine returns an error.
+ *
+ * @param [in]     ep       UCP endpoint that is used for the receive operation.
+ * @param [in]     buffer   Pointer to the buffer to receive the data to.
+ * @param [in]     count    Number of elements to receive into @a buffer.
+ * @param [out]    length   Size of the received data in bytes. The value is
+ *                          valid only if return code is UCS_OK.
+ * @note                    The amount of data received, in bytes, is always an
+ *                          integral multiple of the @a datatype size.
+ * @param [in]     flags    Flags defined in @ref ucp_stream_recv_flags_t.
```
Comment (brminich) [inline src/ucp/api/ucp.h:2785]: same as send, probably no need for flags, need to clarify what parameters are relevant

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -2732,6 +2764,39 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * @param [in]     param    Operation parameters, see @ref ucp_request_param_t
```
Comment (brminich) [inline src/ucp/api/ucp.h:2786]: missing `.`

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h:2469]: yep

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: added

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: flags are removed

File: src/ucp/api/ucp.h (hunk 3 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: flags are removed

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -2449,6 +2449,36 @@ ucs_status_ptr_t ucp_stream_send_nb(ucp_ep_h ep, const void *buffer, size_t coun
...
+ *
+ * @note The user should not modify any part of the @a buffer after this
+ *       operation is called, until the operation completes.
+ *
+ * @param [in]  ep          Destination endpoint handle.
+ * @param [in]  buffer      Pointer to the message buffer (payload).
+ * @param [in]  count       Number of elements to send.
+ * @param [in]  param       Operation parameters, see @ref ucp_request_param_t.
+ *
+ * @return NULL             - The send operation was completed immediately.
```
Comment (brminich) [inline src/ucp/api/ucp.h:2471]: not aligned by `-`

File: src/ucp/api/ucp.h [hunk 6]
```diff
@@ -2449,6 +2449,36 @@ ucs_status_ptr_t ucp_stream_send_nb(ucp_ep_h ep, const void *buffer, size_t coun
...
+ * it is safe to reuse the source @e buffer. If the send operation is
+ * completed immediately the routine returns UCS_OK.
+ *
+ * @note The user should not modify any part of the @a buffer after this
+ *       operation is called, until the operation completes.
+ *
+ * @param [in]  ep          Destination endpoint handle.
+ * @param [in]  buffer      Pointer to the message buffer (payload).
+ * @param [in]  count       Number of elements to send.
+ * @param [in]  param       Operation parameters, see @ref ucp_request_param_t.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:2469]: so the stream flags would also be part of ucp_request_param_t?

File: src/ucp/api/ucp.h [hunk 7]
```diff
@@ -2732,6 +2762,38 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * @return NULL                 - The receive operation was completed
+ *                                immediately.
+ * @return UCS_PTR_IS_ERR(_ptr) - The receive operation failed.
+ * @return otherwise            - Operation was scheduled for receive. A request
+ *                                handle is returned to the application in order.
+ *                                to track progress of the operation.
+ */
+ucs_status_ptr_t ucp_stream_recv_nbx(ucp_ep_h ep, void *buffer, size_t count,
+                                     size_t *length,
+                                     const ucp_request_param_t *param);
```
Comment (yosefe) [inline src/ucp/api/ucp.h:2794]: should we add STREAM_WAITALL to ucp_request_param_t flags?

File: src/ucp/api/ucp.h (hunk 5 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: fixed

File: src/ucp/api/ucp.h (hunk 6 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h:2469]: it could be, but value will not be same as for ```ucp_stream_send_nb``` call

File: src/ucp/api/ucp.h (hunk 7 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h:2794]: I think yes
@evgeny-leksikov any ideas?

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: fixed

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: technically yes, but not sure how it could be used

File: src/ucp/api/ucp.h (hunk 6 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h:2469]: IMO we can define common field `params::flags` and document which flags are relevant for which operation (not only stream), i.e. `ucp_stream_recv_flags_t` is used for `ucp_stream_recv_nbx` and will be ignored for `ucp_stream_send_nbx`.  Just need to document it properly and accurate.

File: src/ucp/api/ucp.h (hunk 7 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h:2794]: yes, we should, see my thoughts here https://github.com/openucx/ucx/pull/4993#discussion_r407339982

Comment (hoopoepg) [reply src/ucp/api/ucp.h:2800]: ok, added

Comment (brminich) [reply src/ucp/api/ucp.h:2800]: also please add `ucp_stream_recv_callback_t ` type to the params (as a union of recv cbs).

Comment (hoopoepg) [reply src/ucp/api/ucp.h:2801]: done: added new callback type

File: src/ucp/api/ucp.h (hunk 6 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h:2470]: right, fixed

File: src/ucp/api/ucp.h [hunk 8]
```diff
@@ -1198,8 +1198,9 @@ typedef struct {
      * send or receive operation is completed.
      */
     union {
-        ucp_send_nbx_callback_t     send;
-        ucp_tag_recv_nbx_callback_t recv;
+        ucp_send_nbx_callback_t        send;
+        ucp_tag_recv_nbx_callback_t    recv;
+        ucp_stream_recv_nbx_callback_t stream;
```
Comment (brminich) [inline src/ucp/api/ucp.h:1203]: `recv_stream?`

File: src/ucp/api/ucp_def.h [hunk 9]
```diff
@@ -433,6 +433,28 @@ typedef void (*ucp_stream_recv_callback_t)(void *request, ucs_status_t status,
...
+ *
+ * @param [in]  request   The completed receive request.
+ * @param [in]  status    Completion status. If the send operation was completed
+ *                        successfully UCS_OK is returned. Otherwise,
+ *                        an @ref ucs_status_t "error status" is returned.
+ * @param [in]  length    The size of the received data in bytes, always
+ *                        boundary of base datatype size. The value is valid
+ *                        only if the status is UCS_OK.
+ * @param [in]  user_data User data passed to "user_data" value,
+ *                        see @ref ucp_request_param_t
```
Comment (brminich) [inline src/ucp/api/ucp_def.h:452]: missing `.` in the end

Comment (hoopoepg) [reply src/ucp/api/ucp_def.h]: added ```.```

File: src/ucp/api/ucp.h (hunk 8 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: renamed to ```recv_stream```

File: src/ucp/api/ucp.h [hunk 10]
```diff
@@ -2774,10 +2774,6 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
  * complete when the message is delivered to the buffer. If the receive
  * operation cannot be started, the routine returns an error.
  *
- * @note This operation supports specific flags, which can be passed
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: need to move this to `@param` doc

Comment (hoopoepg) [reply src/ucp/api/ucp.h]: will not work - flags are added in separate PR
will update wording after https://github.com/openucx/ucx/pull/4984 is merged

Comment (hoopoepg) [reply src/ucp/api/ucp.h]: updated wording

File: src/ucp/api/ucp.h [hunk 11]
```diff
@@ -2824,6 +2855,42 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * complete when the message is delivered to the buffer. If the receive
+ * operation cannot be started, the routine returns an error.
+ *
+ * @param [in]     ep       UCP endpoint that is used for the receive operation.
+ * @param [in]     buffer   Pointer to the buffer to receive the data to.
+ * @param [in]     count    Number of elements to receive into @a buffer.
+ * @param [out]    length   Size of the received data in bytes. The value is
+ *                          valid only if return code is UCS_OK.
+ * @note                    The amount of data received, in bytes, is always an
+ *                          integral multiple of the @a datatype size.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:2875]: move this not after params/return value doc text

File: src/ucp/api/ucp.h [hunk 12]
```diff
@@ -2824,6 +2855,42 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * @param [in]     param    Operation parameters, see @ref ucp_request_param_t.
+ *                          This operation supports specific flags, which can be
+ *                          passed in @a param by @ref ucp_request_param_t.flags.
+ *                          The exact set of flags is defined by
+ *                          @ref ucp_stream_recv_flags_t.
+ *
+ * @return NULL                 - The receive operation was completed
+ *                                immediately.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:2883]: how is length returned in this case?

File: src/ucp/api/ucp.h [hunk 13]
```diff
@@ -2824,6 +2855,42 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * size @a count object on the endpoint @a ep. The routine is non-blocking
+ * and therefore returns immediately. The receive operation is considered
+ * complete when the message is delivered to the buffer. If the receive
+ * operation cannot be started, the routine returns an error.
+ *
+ * @param [in]     ep       UCP endpoint that is used for the receive operation.
+ * @param [in]     buffer   Pointer to the buffer to receive the data to.
+ * @param [in]     count    Number of elements to receive into @a buffer.
+ * @param [out]    length   Size of the received data in bytes. The value is
+ *                          valid only if return code is UCS_OK.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:2873]: UCS_OK -> NULL

Comment (hoopoepg) [reply src/ucp/api/ucp.h]: yep, fixed

File: src/ucp/api/ucp.h (hunk 12 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h:2884]: will not - same as for recv_nbx
to force callback or wait for complete message user should use flags (NO_IMM_CMPL or WAIT_ALL)

File: src/ucp/api/ucp.h (hunk 11 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: moved upper... but I just copy-pasted this text from ucp_stream_recv_nb description - there is same wording/format

Comment (yosefe) [reply src/ucp/api/ucp.h]: i think it should be moved lower (after `@return otherwise`)

File: src/ucp/api/ucp.h (hunk 12 above)
Comment (yosefe) [reply src/ucp/api/ucp.h:2884]: i mean, for case of NULL, we can say '*length' is set to the received length?

File: src/ucp/api/ucp.h (hunk 11 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: ok, moved down

File: src/ucp/api/ucp.h (hunk 12 above)
Comment (hoopoepg) [reply src/ucp/api/ucp.h]: added comment

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls take a look?

**Comment by hoopoepg** [comment]
updated changes

**Comment by hoopoepg** [comment]
@tonycurtis I updated changes. Is it ok for now?

**Comment by hoopoepg** [comment]
@tonycurtis fixed

thank you

**Comment by yosefe** [comment]
@shamisp can you pls take a look?

File: src/ucp/api/ucp.h [hunk 14]
```diff
@@ -2824,6 +2855,45 @@ ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t count,
...
+ * @param [in]     ep       UCP endpoint that is used for the receive operation.
+ * @param [in]     buffer   Pointer to the buffer that will receive the data.
+ * @param [in]     count    Number of elements to receive into @a buffer.
+ * @param [out]    length   Size of the received data in bytes. The value is
+ *                          valid only if return code is NULL.
+ * @param [in]     param    Operation parameters, see @ref ucp_request_param_t.
+ *                          This operation supports specific flags, which can be
+ *                          passed in @a param by @ref ucp_request_param_t.flags.
+ *                          The exact set of flags is defined by
+ *                          @ref ucp_stream_recv_flags_t.
```
Comment (shamisp) [inline src/ucp/api/ucp.h:2878]: Is the callback option or must be set ?

Comment (hoopoepg) [reply src/ucp/api/ucp.h:2878]: it is optional, configured by param op_attr_mask flags

**Comment by shamisp** [comment]
👍

**Comment by yosefe** [comment]
@hoopoepg pls squash

File: src/ucp/stream/stream_send.c [hunk 15]
```diff
@@ -139,6 +139,15 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_send_nb,
...
 }
 
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_send_nbx,
+                 (ep, buffer, count, param),
+                 ucp_ep_h ep, const void *buffer, size_t count,
+                 const ucp_request_param_t *param)
+{
+    return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+}
+
```
Comment (yosefe) [inline src/ucp/stream/stream_send.c:149]: pls remove extra space

Comment (hoopoepg) [reply src/ucp/stream/stream_send.c:149]: done + squashed

**Pull request merged**